### PR TITLE
Send message on callout mutation; subscription; subscriptions improved typings

### PR DESCRIPTION
**Pull request by hero101**
### Describe the background of your pull request

What does your pull request do? Does it solve a bug (which one?), add a feature?

### Additional context

Add any other context

### Governance 

- [ ] Documentation is added
- [ ] Test cases are added or updated

By submitting this pull request I confirm that:
- I've read the contributing document https://github.com/alkem-io/.github/blob/master/CONTRIBUTING.md.
- I've read and understand the Code of Conduct https://github.com/alkem-io/.github/blob/master/CODE_OF_CONDUCT.md.
- I understand that any contributions or suggestions I made may make it into the actual code. I've read the License 
     https://github.com/alkem-io/Coordination/blob/master/LICENSE.
- I understand that submitting the pull request requires agreeing to and signing the contributor license agreement.

### CODE REVIEW

File: src/common/decorators/subscription.decorator.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { IncomingHttpHeaders } from 'http';
+import {
+  ReturnTypeFunc,
+  Subscription,
+  SubscriptionOptions,
+} from '@nestjs/graphql';
+import { GraphQLResolveInfo } from 'graphql';
+import { AgentInfo } from '@src/core';
```
Comment (techsmyth) [inline src/common/decorators/subscription.decorator.ts]: Import should be more specific - this is pulling too much in, especially for a class that will be widely used.

File: src/common/decorators/subscription.decorator.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+export interface SubscriptionResolveContext {
```
Comment (techsmyth) [inline src/common/decorators/subscription.decorator.ts]: one interface / type per file is standard across the server, please split out into a separate file

File: src/common/decorators/subscription.decorator.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  req: {
+    authInfo: unknown;
+    headers: IncomingHttpHeaders;
+    user: AgentInfo;
+  };
+}
+
+export interface TypedSubscriptionOptions<TPayload, TVariables, TContext>
```
Comment (techsmyth) [inline src/common/decorators/subscription.decorator.ts]: ditto into a separate file

File: src/common/decorators/subscription.decorator.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  resolve: (
+    payload: TPayload,
+    args: TVariables,
+    context: TContext,
+    info: GraphQLResolveInfo
+  ) => TPayload | Promise<TPayload>; // todo: change to provide other return type than payload
+}
+
+export function TypedSubscription<
```
Comment (techsmyth) [inline src/common/decorators/subscription.decorator.ts]: file name should match the decorator name

File: src/domain/collaboration/aspect/aspect.resolver.subscriptions.ts [hunk 5]
```diff
@@ -28,60 +29,52 @@ export class AspectResolverSubscriptions {
...
   async aspectCommentsMessageReceived(
     @CurrentUser() agentInfo: AgentInfo,
-    @Args({
-      name: 'aspectID',
-      type: () => UUID,
-      description: 'The ID of the Aspect to subscribe to.',
-      nullable: false,
-    })
-    aspectID: string
+    @Args({ nullable: false }) args: AspectMessageReceivedArgs
```
Comment (techsmyth) [inline src/domain/collaboration/aspect/aspect.resolver.subscriptions.ts:68]: The type name here needs consideration. I do not like args. AspectMessageReceivedSubscriptionArgs? Or we switch to using an input DTO. In the case of every mutation we have a DTO to wrap all arguments.

File: src/domain/collaboration/callout/callout.resolver.mutations.ts [hunk 6]
```diff
@@ -61,6 +75,63 @@ export class CalloutResolverMutations {
     return await this.calloutService.deleteCallout(deleteData.ID);
   }
 
+  @UseGuards(GraphqlGuard)
+  @Mutation(() => CommunicationMessageResult, {
+    description: 'Send a message on a Comments Callout',
+  })
+  @Profiling.api
+  async sendMessageOnCallout(
```
Comment (techsmyth) [inline src/domain/collaboration/callout/callout.resolver.mutations.ts]: Interesting that you have a seprate method for this, as in theory it should be possible to achieve the same effect with sending a message on a Comments directly. To discuss, but I think on reflection this makes sense. There is a strong argument then that the mutation to send a message directly on a Comments entity should be removed as we will need a similar mutation on the Aspect entity for sending a message.

File: src/domain/collaboration/callout/callout.resolver.subscriptions.ts [hunk 7]
```diff
@@ -92,4 +94,29 @@ export class CalloutResolverSubscriptions {
       SubscriptionType.CALLOUT_ASPECT_CREATED
     );
   }
+
+  @UseGuards(GraphqlGuard)
+  @TypedSubscription<CalloutMessageReceived, CalloutMessageReceivedArgs>(
```
Comment (techsmyth) [inline src/domain/collaboration/callout/callout.resolver.subscriptions.ts]: Should the type for the Payload not have Payload then in the name, similar to Args?

File: src/domain/collaboration/callout/callout.resolver.subscriptions.ts [hunk 8]
```diff
@@ -92,4 +94,29 @@ export class CalloutResolverSubscriptions {
...
+      },
+    }
+  )
+  async calloutMessageReceived(
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    @CurrentUser() agentInfo: AgentInfo,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    @Args({ nullable: false }) args: CalloutMessageReceivedArgs
+  ) {
+    // todo
```
Comment (techsmyth) [inline src/domain/collaboration/callout/callout.resolver.subscriptions.ts]: todo?

File: src/domain/collaboration/callout/dto/callout.dto.event.message.received.ts [hunk 9]
```diff
@@ -1,8 +1,8 @@
 import { CommunicationMessageResult } from '@domain/communication/message/communication.dto.message.result';
 import { Field, ObjectType } from '@nestjs/graphql';
 
-@ObjectType('CalloutCommentsMessageReceived')
-export class CalloutCommentsMessageReceived {
+@ObjectType('CalloutMessageReceived')
+export class CalloutMessageReceived {
```
Comment (techsmyth) [inline src/domain/collaboration/callout/dto/callout.dto.event.message.received.ts:5]: Payload / result?

File: src/domain/collaboration/callout/dto/callout.dto.event.message.received.ts [hunk 10]
```diff
@@ -12,6 +12,12 @@ export class CalloutCommentsMessageReceived {
   })
   calloutID!: string;
 
+  @Field(() => String, {
```
Comment (techsmyth) [inline src/domain/collaboration/callout/dto/callout.dto.event.message.received.ts]: This should not be needed if sending a message to a Callout as there is exactly on Comments associated with a Callout, but might make the client code easier?

File: src/domain/communication/communication/dto/communication.discussion.updated.args.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { ArgsType, Field } from '@nestjs/graphql';
+import { UUID } from '@domain/common/scalars';
+
+@ArgsType()
+export class CommunicationDiscussionUpdatedArgs {
+  @Field(() => UUID, {
+    description: 'The Communication to receive Updates from.',
```
Comment (techsmyth) [inline src/domain/communication/communication/dto/communication.discussion.updated.args.ts]: receive Updates?

**Review by techsmyth — CHANGES_REQUESTED** [review]
Very interesting PR, and nice to see the typing coming through. 

It does raise for me two points that need deeper discussion:
- we need to probably separate between the Payload that is received by a subscription and the result returned to the user. At the moment we are reusing the same type which is not a good practice. 
- we should have that all mutations **and** subscriptions use input DTOs. This PR adds usage of ArgsType, but given we use input DTOs everywhere else we also should use them on Subscriptions. 

Nicely done.

File: src/domain/collaboration/callout/dto/callout.dto.event.message.received.ts (hunk 9 above)
Comment (techsmyth) [reply src/domain/collaboration/callout/dto/callout.dto.event.message.received.ts:5]: Fine if we split the payload out

File: src/domain/communication/communication/dto/communication.discussion.updated.args.ts (hunk 11 above)
Comment (hero101) [reply src/domain/communication/communication/dto/communication.discussion.updated.args.ts:7]: Please format this into a suggestion. I am not sure what are you asking about or suggesting.

File: src/domain/collaboration/callout/callout.resolver.subscriptions.ts (hunk 8 above)
Comment (hero101) [reply src/domain/collaboration/callout/callout.resolver.subscriptions.ts]: The whole body of the initial subscription is missing - the authorization, and the logging.
Also, the filter and resolve methods are not implemented (forgotten).

File: src/domain/collaboration/aspect/aspect.resolver.subscriptions.ts (hunk 5 above)
Comment (techsmyth) [reply src/domain/collaboration/aspect/aspect.resolver.subscriptions.ts:68]: Not an issue for this PR, we need a broader discussion on the args to use for Subscriptions

**Review by techsmyth — APPROVED** [review]

**Pull request merged**